### ISSUE TITLE
[enhancement] introduce enumerate in for loops

### ISSUE THREAD

**Issue by baloan** [KSP-KOS/KOS#1017]
In for loops often you require the index and the value of a list element. In kOS it is:
```
set i to ps:iterator.
until not i:next {
    set p to i:value.
    ...
}
```
Instead it could be (shamelessly inspired by python)
```
for i,p in enumerate(ps) {
    ...
}
```

**Comment by abenkovskii** [comment]
I'm for OOP style:
```
for kv_pair in list:enumerate {
    print kv_pair[0] + " " + kv_pair[1].
}
```

**Comment by Dunbaratu** [comment]
So if I understand this, the implementation would simply be that ``:enumerate`` would construct a newly made list out of the existing enumerable type, in which the new list is a list of lists, each element of which is a pair where [0] is the key (or index number) and [1] is the value at that key.

I'd almost rather see an explicit keyval pair type, to eliminate the [0] and [1].  have it be a list of KeyValPair where KeyValPair has a ``:key`` and ``:value`` suffix.

**Comment by erendrake** [comment]
I agree that the `pair` structure would also be useful for the associated array, when i finally get it finished :P

**Comment by abenkovskii** [comment]
I like the pair idea.

**Comment by Dunbaratu** [comment]
If done carefully, you can remove the expense of copying the values of the original list by just having the pair objects not *really* contain `:value`' in them, bu instead have the pair's `:value` be a method that gets he value live, from the key.

**Comment by Dunbaratu** [comment]
@erendrake, is this issue dead now with our new types or is it still alive?
